src/client/Microsoft.Identity.Client/Http/HttpManagerWithRetry.cs

```diff
@@ -0,0 +1,195 @@
+            IDictionary<string, string> headers,
+            StringContent body,
+            ILoggerAdapter logger,
+            CancellationToken cancellationToken = default)
+        {
+            return await SendRequestAsync(uri, headers, body, HttpMethod.Post, logger, retry: true, doNotThrow: true, cancellationToken: cancellationToken).ConfigureAwait(false);
+        }
+
+        protected override HttpClient GetHttpClient()
+        {
+            return _httpClientFactory.GetHttpClient();
+        }
+
+        protected override async Task<HttpResponse> SendRequestAsync(
+            Uri endpoint,
+            IDictionary<string, string> headers,
+            HttpContent body,
+            HttpMethod method,
+            ILoggerAdapter logger,
+            bool doNotThrow = false,
+            bool retry = true,
+            CancellationToken cancellationToken = default)
+        {
+            Exception timeoutException = null;
+            bool isRetriableStatusCode = false;
+            HttpResponse response = null;
+            bool isRetriable;
+            
+            try
+            {
+                HttpContent clonedBody = body;
+                if (body != null)
+                {
+                    // Since HttpContent would be disposed by underlying client.SendAsync(),
+                    // we duplicate it so that we will have a copy in case we would need to retry
+                    clonedBody = await CloneHttpContentAsync(body).ConfigureAwait(false);
+                }
+
+                using (logger.LogBlockDuration("[HttpManager] ExecuteAsync"))
+                {
+                    response = await ExecuteAsync(endpoint, headers, clonedBody, method, logger, cancellationToken).ConfigureAwait(false);
+                }
+
+                if (response.StatusCode == HttpStatusCode.OK)
+                {
+                    return response;
+                }
+
+                logger.Info(() => string.Format(CultureInfo.InvariantCulture,
+                    MsalErrorMessage.HttpRequestUnsuccessful,
+                    (int)response.StatusCode, response.StatusCode));
+
+                isRetriableStatusCode = IsRetryableStatusCode((int)response.StatusCode);
+                isRetriable = isRetriableStatusCode && !HasRetryAfterHeader(response);
+            }
+            catch (TaskCanceledException exception)
+            {
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    logger.Info("The HTTP request was cancelled. ");
+                    throw;
+                }
+
+                logger.Error("The HTTP request failed. " + exception.Message);
+                isRetriable = true;
+                timeoutException = exception;
+            }
+
+            if (isRetriable && retry)
+            {
+                logger.Info("Retrying one more time..");
+                await Task.Delay(TimeSpan.FromSeconds(1)).ConfigureAwait(false);
```

Comment:
suggested wait time is 5 seconds : https://learn.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific#retry-usage-guidance

Reply:
But if this is specific to MI, then should we not do the Exponential back-off we discussed? 1 2 4 ?

Reply:
Discussed this with @bgavrilMS , we will not implement the retry logic wih exponential back-off since that is a functionality provided by Azure SDK for all the calls. It is by default set to 3 retries but user can configure and change it. MSAL is a low level library and we are not planning to implement all that retry logic. We will instead return IsRetriable true and document that users can choose to implement the retry.